IntelliTect.TestTools.Selenate/IntelliTect.TestTools.Selenate/ElementsHandler.cs

```diff
@@ -98,21 +111,52 @@ public bool ContainsText(string text)
         /// </summary>
         /// <param name="predicate">The criteria to attempt to match on.</param>
         /// <returns></returns>
-        public IWebElement GetSingleExistingElement(Func<IWebElement, bool> predicate)
+        public IWebElement GetSingleWebElement(Func<IWebElement, bool> predicate)
+        {
+            IList<IWebElement> elems = GetElements(predicate);
+
+            if (elems.Count != 1)
+            {
+                throw new ArgumentOutOfRangeException(nameof(predicate), "The provided predicate did not match exactly one result.");
+            }
+
+            return elems[0];
+        }
+
+        /// <summary>
+        /// Gets all elements found by <see cref="Locator"/>, matching a given predicate.
+        /// </summary>
+        /// <param name="predicate">The function used to filter to one or more IWebElements</param>
+        /// <returns>A list of found IWebElements</returns>
+        /// <exception cref="NoSuchElementException"></exception>
+        /// <exception cref="ArgumentOutOfRangeException"></exception>
+        public IList<IWebElement> GetAllWebElements(Func<IWebElement, bool> predicate)
+        {
+            return GetElements(predicate);
+
+        }
+
+        public IList<IWebElement> GetElements(Func<IWebElement, bool> predicate)
         {
             IWait<IWebDriver> wait = Wait;
             wait.IgnoreExceptionTypes(typeof(NoSuchElementException));
-            return wait.Until(d =>
+
+#pragma warning disable CS8603 // Possible null reference return. Needed for Selenium retry.
+            return wait.Until(_ =>
             {
-                var foundElems = d.FindElements(Locator);
-                if (foundElems is null || foundElems.Count == 0) throw new NoSuchElementException($"No element found matching pattern: {Locator}");
-                var foundElem = foundElems.Where(predicate).ToList();
-                if (foundElem.Count != 1)
+                IReadOnlyCollection<IWebElement> foundElems = SearchContext.FindElements(Locator);
+                if (foundElems is null || foundElems.Count == 0) throw new NoSuchElementException($"No elements found matching pattern: {Locator}");
```

Comment:
The variable is declared non-nullable, so the null check should not be needed, or the variable type should be changed.
```suggestion
                if (foundElems.Count == 0) throw new NoSuchElementException($"No elements found matching pattern: {Locator}");
```

Reply:
I'll double check Selenium's default behavior and change as appropriate.

Reply:
Your suggested change is correct. :) Selenium returns an empty collection.